src/main/scala/lantern/dslapi.scala

```diff
@@ -442,13 +451,261 @@ int main(int argc, char *argv[]) {
+    "CUDA_CALL(cudaMallocManaged(&" + buffer + ", " + count + " * sizeof(" + memType + ")));"
+  }
+
+  override def emitNode(sym: Sym[Any], rhs: Def[Any]) = rhs match {
+    case a@ArrayNew(n) =>
+      // Unified CPU/GPU memory via `cudaMallocManaged` is more convenient than `cudaMalloc`, but less performant.
+      // We can use a similar memory pool technique with `cudaMallocManaged`.
+      val arrType = remap(a.m)
+      stream.println(arrType + "* " + quote(sym) + "; " + getCudaMallocManagedString(quote(sym), quote(n), arrType))
+      // stream.println(arrType + "* " + quote(sym) + "; " + getCudaMallocString(quote(sym), quote(n), arrType))
+    case _ => super.emitNode(sym,rhs)
+  }
+
+  override def emitSource[A:Typ](args: List[Sym[_]], body: Block[A], functionName: String, out: java.io.PrintWriter) = {
+    withStream(out) {
+      stream.println("""#include <assert.h>
+#include <err.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <functional>
+#include <iostream>
+#include <math.h>
+#include <memory>
+#include <random>
+#include <stdint.h>
+#include <stdio.h>
+#include <sys/mman.h>
+#include <sys/stat.h>
+#include <sys/time.h>
+#include <time.h>
+#include <unistd.h>
+
+#include <cuda_runtime.h>
+#include "cublas_v2.h"
+
+using namespace std;
+
+#ifndef MAP_FILE
+#define MAP_FILE MAP_SHARED
+#endif
+
+#define CUDA_CALL(f) { \
+  cudaError_t err = (f); \
+  if (err != cudaSuccess) { \
+    std::cerr << "Error occurred: " << err << std::endl; \
+    std::exit(1); \
+  } \
+}
+
+#define CUBLAS_CALL(f) { \
+  cublasStatus_t stat = (f); \
+  if (stat != CUBLAS_STATUS_SUCCESS) { \
+    std::cerr << "Error occurred: " << err << std::endl; \
+    exit(1); \
+  } \
+}
+
+int fsize(int fd) {
+  struct stat stat;
+  int res = fstat(fd, &stat);
+  return stat.st_size;
+}
+
+int printll(char *s) {
+  while (*s != '\n' && *s != ',' && *s != '\t') {
+    putchar(*s++);
+  }
+  return 0;
+}
+
+long hash(char *str0, int len) {
+  unsigned char *str = (unsigned char *)str0;
+  unsigned long hash = 5381;
+  int c;
+
+  while ((c = *str++) && len--)
+    hash = ((hash << 5) + hash) + c; /* hash * 33 + c */
+
+  return hash;
+}
+
+int HEAP_SIZE = 1073741826; // 1048576; // 2147483652; // 536870912; // 268435456; // 2097152;
+void *mallocBase = malloc(HEAP_SIZE);
+void *mallocAddr = mallocBase;
+void *waterMark = mallocBase;
+void *myMalloc(size_t bytes) {
+  void *res = mallocAddr;
+  mallocAddr = (void *)((char *)mallocAddr + bytes);
+  return res;
+}
+
+int timeval_subtract(struct timeval *result, struct timeval *t2, struct timeval *t1) {
+  long int diff = (t2->tv_usec + 1000000 * t2->tv_sec) - (t1->tv_usec + 1000000 * t1->tv_sec);
+  result->tv_sec = diff / 1000000;
+  result->tv_usec = diff % 1000000;
+  return (diff < 0);
+}
+
+void Snippet(char *);
+
+std::random_device rd{};
+std::mt19937 gen{rd()};
+std::normal_distribution<> d{0, 1};
+cublasHandle_t handle;
+
+int main(int argc, char *argv[]) {
+  CUBLAS_CALL(cublasCreate(&handle));
```

Comment:
I was thinking explicit `init` and `cleanup` methods in the backend interface.

Reply:
I wonder how to hook up `init`/`cleanup` `Backend`-defined methods with the codegen template. AFAICT, the `DslDriverXXX` trait (which contains `DslGenXXX`) and `TensorExp` trait (which contains `Backend`) are decoupled, so I'm not sure how to invoke `Backend`-defined methods in `DslGenXXX`, where codegen templates are defined.

Perhaps I'm missing something, do you have some idea on how to hook up things?

Reply:
I'm sorry, I'm still a bit confused. Do you mean something like this?
```scala
val foo = new DslDriverCublas[String, Unit] with TensorExp {
  backend = new BackendCublas

  @virtualize
  def snippet(x: Rep[String]): Rep[Unit] = {
    backend.init()
    // ...
    Tensor.assertEqual(...)
    backend.cleanup()
  }
}
```

If so, I don't quite believe this is the best approach. I would argue backend setup/cleanup are implementation details that should be hidden from DSL users if possible. Also, if `backend.init()` initializes `cublasHandle_t handle`, then it really should be performed before any computation in the snippet. Burdening DSL users with that responsibility is suboptimal.

Reply:
I wasn't necessarily thinking users should call these, but for example it could be ideally hidden in a `withGPU { ... }` construct or even as part of the DSLDriver class before and after invoking `snippet`.

Reply:
In general, my point here is that codegen templates should be as minimal as possible, and we should define the application logic as far as possible using metaprogramming.

Reply:
I see, that makes sense - thanks for the clarification.

Along those lines, a straightforward idea is to create a function wrapping `snippet` and compiling that function during codegen:
```scala
// `wrapper` could use a better name.
def wrapper(x: Rep[String]): Rep[Unit] = {
  backend.init()
  val result = snippet(x)
  backend.cleanup()
  return result
}

// During code generation:
lazy val code: String = {
  val source = new java.io.StringWriter()
  // Compile wrapper function.
  codegen.emitSource(wrapper, "Snippet", new java.io.PrintWriter(source))
  source.toString
}
```

If this seems reasonable (and it works), I'll implement it.

Reply:
That looks good, yes. Let's not worry about idempotency at this point.